---
title: "GRNF Guide"
author: "Ryne Ramaker"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

---
title: "GRNF_Vignette"
author: "Ryne Ramaker"
date: "May 9, 2017"
output: html_document
---

###Introduction

The primary function of this package is to provide tools for assessing which genes a DNA associated protien (DAP) likely regulates based on its genome occupancy. The primary function *GeneToPeakDist* computes the distance from gene transcription start sites (TSS) to the nearest DAP binding site. This function can also calculate weighted distances based on chromatin structure information such as topologically associated domains or promoter interactions. The second function *FindPathwayEnrichment* will assess enrichment for proximal binding to Reactome pathway gene sets.

###Installation

There are multiple ways to install this package. The first involves downloading the binary source file from my github page (https://github.com/rramaker/GRNF) and installing the package manually:

```{r, eval=FALSE}
install.packages("/PathToDownloadedFile/GRNF_0.1.0.tgz")
```


The package can be directly installed from github with the devtools package

```{r, eval=FALSE}
library(devtools)
install_github("rramaker/GRNF")
```


Alternatively the package can be downloaded from the CRAN repository. ##Coming Soon!

```{r, eval=FALSE}
install.packages("GRNF")
```

Once installed the functions from GRNF can be made available in your current working environment with the library() command:

```{r}
library(GRNF)
```

###Formatting data

##ChIP-seq peak input (BED file)

The primary required input is a dataframe in BED format (https://genome.ucsc.edu/FAQ/FAQformat.html#format1) with each row representing the genomic location of a unique ChIP-seq peak. There are only three columns. The first column should indicate the chromsome of a binding site. The second and third column should indicate the start and stop position of each binding site. Additional columns after these first three are acceptalbe. An example publically available ChIP-seq data set^1^ has been provided with the package and can be accessed as shown below:

```{r}
data("GM12878_BATF_ChIP")
head(GM12878_BATF_ChIP[,1:3])
```

##Gene position input (GTF file)

The second required input is a dataframe in GTF format (http://www.ensembl.org/info/website/upload/gff.html) with each row representing the genomic location of an annotated gene. There are specifically five required columns that will queried. The first column should indicate the chromosome position of each gene. The format of this column should match the chromosome column of the ChIP-seq peak input described above (i.e. "chr1" vs. "1"). The third and fourth columns should indicate the start and stop position of each gene. The seventh column should contain strand information and the ninth column should be an "attribute" column that at minimum contains the ENSEMBL ID of each gene. An example hg19/GRCh37 ENSEMBL GTF file has been provided with the package and can be accessed as shown below:

```{r}
data("Homo_sapiens.GRCh37.82.chr.gtf")
head(Homo_sapiens.GRCh37.82.chr.gtf)
```

##Topological Associated Domain (TAD) boundary input

Optionally, TAD boundary information can be provided and used to penalize TSS to binding site distances. The required format for TAD boundaries is a dataframe with each row representing a unique TAD. The first column should indicate the chromosome  of each TAD (formatted identically to the ChIP-seq BED and gene GTF inputs) and the second and third columns should indicate the TAD boundaries. An example publically available TAD file^2^ has been provided with the package and can be accessed as shown below:

```{r}
data("GM12878_TAD")
head(GM12878_TAD)
```

##Promoter interaction input

Optionally, information on which genomic regions interact with gene promoters, as measured by assays such as promoter capture Hi-C (PCHC), can be provided to appropriatley reduce TSS to binding site distances if binding sites are found in strong promoter interacting regions. The required input format consists of five columns. The first three columns should indicate the chromsome (formatted consistently with ChIP-seq BED files and gene GTF files described above), start and stop positions of each "captured" promoter interacting region. The fourth column should provide a metric associated with the interaction frequency of each "captured" region with it's respective promoter "bait". The fifth column should contain the ENSEMBL ID of the gene who's promoter acts as "bait" for an interaction. An example input describing publically available promoter interaction data has been provided with the package and can be accessed as shown below:

```{r}
data("GM12878_PCHC")
head(GM12878_PCHC)
```

###Finding ChIP-seq peak to gene TSS distances with GeneToPeakDist


###Performing pathway enrichment analysis for proximal ChIP-seq peaks
